containers: Various maintenance tasks

foersleo opened the pull request:
*Issue #, if available:* #35 #38 

*Description of changes:*

* Update `aws-nitro-enclaves-sdk-c` to latest version v0.4.3
* fix copying from fill_image (#35)
* Get base images from ECR instead of DockerHub
* Pick up osx compatibility from old PR (#32 , fixes #38 )
* Update k8s version to be compatible with latest EKS


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

dpdornseifer committed:
made script OSX compatible

Commit:
container: kms: Update aws-nitro-enclaves-sdk-c to v0.4.3

Signed-off-by: Leonard Foerster <foersleo@amazon.com>

Commit:
containers: Move individual subdirectories from full_image

Newer docker (>=23.0) uses BuildKit as the default builder. That behaves
differently when copying subdirectories to symlinks on the target, and
will bail out with error message

"ERROR: failed to solve: cannot copy to non-directory: ..."

Breaking the copying up into individual subdirectories fixes this issue
as it eliminates ambiguity in the order of operations that mey influence
each other due to the symbolic links.

Signed-off-by: Leonard Foerster <foersleo@amazon.com>

Commit:
container: Get base images from ECR instead of DockerHub

The images are semantically identical, so this is not a functional
change.

Signed-off-by: Leonard Foerster <foersleo@amazon.com>

clupuishere commented on container/hello/Dockerfile:
```diff
@@ -41,7 +41,8 @@ FROM amazonlinux:2 as image
 # to shrink the final image size.
 #
-COPY --from=full_image /ne-deps /
+RUN mkdir p ne-deps
```
I think we're missing a dash here, i.e. `-p`.

foersleo replied:
Ouch, true.

foersleo commented:
Above force push adds the necessary dash in `containers/hello/Dockerfile` as pointed out by clupuishere.

eugkoira commented on container/hello/Dockerfile:
```diff
@@ -41,7 +41,8 @@ FROM amazonlinux:2 as image
 #
-COPY --from=full_image /ne-deps /
+RUN mkdir -p ne-deps
+COPY --from=full_image /ne-deps /ne-deps
```
Sorry, but I don't understand how this should work.
We've created a shrinked root fs in `/ne-deps` with assumption to copy it all together to a target container root folder. IIRC it [didn't work](https://github.com/aws/aws-nitro-enclaves-with-k8s/issues/35) because something changed in how `/lib64` is represented (now it's a link) in the latest AL.

But now if we copy `/ne-deps` to the target `/ne-deps` how all the binaries are found if you just call `nitro-cli` in `run.sh`?

foersleo replied:
I believe what you are talking about is happening earlier and we are in fact making a deep copy of necessary libraries in lines 20-26 by using the `-L` option in the copying.

foersleo replied:
I actually had gotten that wrong.

Splitting up the copy into separate sub-directories of the source does the trick of helping buildkit over copying into symlinked directories.

Commit:
settings: update to k8s version 1.32

Update to k8s version 1.32 to be compatible with latest EKS.
Starting with version 1.30 EKS started to use AL2023 instead of AL2 as
the default ami for nodes. On AL2023 `aws-nitro-enclaves-cli` is not
part of the extras repository anymore, but can be installed from the
core repositories.

While here, move the dependency installation into cloud-config section.

This commit is partly based on a previous PR by dornseid@amazon.com [1]

[1] https://github.com/aws/aws-nitro-enclaves-with-k8s/pull/32

Signed-off-by: Leonard Foerster <foersleo@amazon.com>
